File: app/src/main/java/com/elvotra/clean/presentation/ui/fragments/PostsListFragment.java
package com.elvotra.clean.presentation.ui.fragments;

import android.content.Intent;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.content.ContextCompat;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.elvotra.clean.R;
import com.elvotra.clean.presentation.contract.PostsContract;
import com.elvotra.clean.presentation.model.PostViewItem;
import com.elvotra.clean.presentation.ui.activities.PostDetailsActivity;
import com.elvotra.clean.presentation.ui.adapters.PostsRecyclerAdapter;
import com.elvotra.clean.presentation.ui.widgets.ScrollChildSwipeRefreshLayout;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

import static com.elvotra.clean.presentation.ui.activities.PostDetailsActivity.POST_ID;

public class PostsListFragment extends Fragment implements PostsContract.View {

    private PostsContract.IPostsPresenter postsPresenter;

    private PostsRecyclerAdapter postsRecyclerAdapter;

    @BindView(R.id.fragment_posts_list_recycler_view)
    RecyclerView postsRecyclerView;
    @BindView(R.id.fragment_posts_list_message)
    TextView errorMessage;
    @BindView(R.id.refresh_layout)
    ScrollChildSwipeRefreshLayout swipeRefreshLayout;

    public PostsListFragment() {
    }

    public static PostsListFragment newInstance() {
        PostsListFragment fragment = new PostsListFragment();
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_posts_list, container, false);

        ButterKnife.bind(this, rootView);

        RecyclerView.LayoutManager mLayoutManager = new LinearLayoutManager(getActivity());
        postsRecyclerView.setLayoutManager(mLayoutManager);

        swipeRefreshLayout.setColorSchemeColors(
                ContextCompat.getColor(getActivity(), R.color.colorPrimary),
                ContextCompat.getColor(getActivity(), R.color.colorAccent),
                ContextCompat.getColor(getActivity(), R.color.colorPrimaryDark)
        );
        swipeRefreshLayout.setScrollUpChild(postsRecyclerView);

        swipeRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                postsPresenter.loadPosts(true);
            }
        });
        postsPresenter.loadPosts(true);

        return rootView;

    }

    @Override
    public void onResume() {
        super.onResume();
    }

    @Override
    public void setPresenter(PostsContract.IPostsPresenter presenter) {
        postsPresenter = presenter;
    }

    @Override
    public void showPostsList(List<PostViewItem> postViewItems) {
        errorMessage.setVisibility(View.GONE);

        postsRecyclerView.setVisibility(View.VISIBLE);

        postsRecyclerAdapter = new PostsRecyclerAdapter(getContext(), postViewItems,
                new PostsRecyclerAdapter.PostsListItemClickListener() {

                    @Override
                    public void onPostClicked(int postId) {
                        postsPresenter.openPostDetails(postId);
                    }

                });

        postsRecyclerView.setAdapter(postsRecyclerAdapter);
    }

    @Override
    public void showPostDetails(int postId) {
        Intent i = new Intent(getActivity(), PostDetailsActivity.class);
        i.putExtra(POST_ID, postId);
        startActivity(i);
    }

    @Override
    public void showNoResults() {
        postsRecyclerView.setVisibility(View.GONE);
        errorMessage.setVisibility(View.VISIBLE);
        errorMessage.setText(getString(R.string.no_posts));
    }

    @Override
    public void showProgress() {
        swipeRefreshLayout.setRefreshing(true);
    }

    @Override
    public void hideProgress() {
        swipeRefreshLayout.setRefreshing(false);
    }

    @Override
    public void showError(String message) {
        postsRecyclerView.setVisibility(View.GONE);
        errorMessage.setVisibility(View.VISIBLE);
        errorMessage.setText(message);
    }

    @Override
    public boolean isActive() {
        return isAdded();
    }
}

File: app/src/main/java/com/elvotra/clean/presentation/ui/fragments/PostsListFragment.kt
package com.elvotra.clean.presentation.ui.fragments

import android.content.Intent
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.content.ContextCompat
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.elvotra.clean.R
import com.elvotra.clean.presentation.contract.PostsContract
import com.elvotra.clean.presentation.model.PostViewItem
import com.elvotra.clean.presentation.ui.activities.PostDetailsActivity
import com.elvotra.clean.presentation.ui.activities.PostDetailsActivity.Companion.POST_ID
import com.elvotra.clean.presentation.ui.adapters.PostsRecyclerAdapter
import com.elvotra.clean.presentation.ui.widgets.ScrollChildSwipeRefreshLayout

class PostsListFragment : Fragment(), PostsContract.View {

    override lateinit var presenter: PostsContract.IPostsPresenter

    private lateinit var postsRecyclerAdapter: PostsRecyclerAdapter

    private lateinit var fragment_posts_list_recycler_view: RecyclerView
    private lateinit var fragment_posts_list_message: TextView
    private lateinit var refresh_layout: ScrollChildSwipeRefreshLayout

    override val isActive: Boolean
        get() = isAdded

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val rootView = inflater.inflate(R.layout.fragment_posts_list, container, false)

        val mLayoutManager = LinearLayoutManager(activity)
        fragment_posts_list_recycler_view.layoutManager = mLayoutManager

        refresh_layout.setColorSchemeColors(
                ContextCompat.getColor(activity!!, R.color.colorPrimary),
                ContextCompat.getColor(activity!!, R.color.colorAccent),
                ContextCompat.getColor(activity!!, R.color.colorPrimaryDark)
        )
        refresh_layout.setScrollUpChild(fragment_posts_list_recycler_view)

        refresh_layout.setOnRefreshListener { presenter.loadPosts(true) }
        presenter.loadPosts(true)

        return rootView

    }

    fun getPresenter(presenter: PostsContract.IPostsPresenter): PostsContract.IPostsPresenter? {
        return this.presenter
    }

    override fun showPostsList(postViewItems: List<PostViewItem>) {
        fragment_posts_list_message.visibility = View.GONE

        fragment_posts_list_recycler_view.visibility = View.VISIBLE

        postsRecyclerAdapter = PostsRecyclerAdapter(context!!, postViewItems, PostsRecyclerAdapter.PostsListItemClickListener { override fun onPostClicked(postId: Int){
            presenter.openPostDetails(postId)
            }
        })

        fragment_posts_list_recycler_view.adapter = postsRecyclerAdapter
        //{ postId: Int -> presenter.openPostDetails(postId) }
    }

    override fun showPostDetails(postId: Int) {
        val i = Intent(activity, PostDetailsActivity::class.java)
        i.putExtra(POST_ID, postId)
        startActivity(i)
    }

    override fun showNoResults() {
        fragment_posts_list_recycler_view.visibility = View.GONE
        fragment_posts_list_message.visibility = View.VISIBLE
        fragment_posts_list_message.text = getString(R.string.no_posts)
    }

    override fun showProgress() {
        refresh_layout.isRefreshing = true
    }

    override fun hideProgress() {
        refresh_layout.isRefreshing = false
    }

    override fun showError(message: String) {
        fragment_posts_list_recycler_view!!.visibility = View.GONE
        fragment_posts_list_message.visibility = View.VISIBLE
        fragment_posts_list_message.text = message
    }

    companion object {

        fun newInstance(): PostsListFragment {
            return PostsListFragment()
        }
    }
}
